Searching box

MieczyslawFrontDev opened the pull request:
Hi !

I'm adding an update to the application where I handled the movie and actors search. To search for a movie I used the filter function which argument is title property. To filter actor I used name argument. No results are handled with the NoResultsPage component. 
After going into the movie details, the search gets the disabled attribute. 
Additionally I've changed name of colors in the theme file and I used for that Name that color generator.

paulgrym reviewed (changes requested):
Works fine! 😀  But we need to use this API https://developers.themoviedb.org/3/search/search-movies to search in all movies, not only in the popular ones :)

paulgrym commented on src/features/APIdata.js:
```diff
@@ -2,6 +2,8 @@ export const APIUrl = "https://api.themoviedb.org/3"
 
 export const APIKey = "951446c9e47a99cb6ccfb776c60023c7"
 
+export const APISearchMoviesUrl = `${APIUrl}/search/movie?api_key=${APIKey}&query=`
```
Maybe something similar to APImovieDetailsUrl function would be better, because in the future we will be passing not only query, but page as well :) 
``` 
export const APImovieSearchUrl = (query) =>
`${APIUrl}/search/movie?api_key=${APIKey}&query=${query}`

paulgrym commented on src/features/movies/MovieList/index.js:
```diff
@@ -1,55 +1,77 @@
-      }
-    </Section>
+      <Section>
+        {query && error || totalResults === 0
```
You fixed a bug in displaying the ResultPage when it shouldn't, but now error handling doesn't work :(.

paulgrym replied:
We don't want NoResultPage to be displayed when there is an error - we want the ErrorPage

paulgrym commented on src/features/movies/MovieList/popularMoviesSaga.js:
```diff
@@ -1,22 +1,37 @@
+} from "./popularMoviesSlice";
 
-function* fetchPopularMoviesWorker() {
+function* fetchPopularMoviesWorker({ payload: { query } }) {
```
We can remove "popular" part of functions names :)

paulgrym commented on src/features/movies/MovieList/popularMoviesSaga.js:
```diff
@@ -1,22 +1,37 @@
     yield delay(1000);
-    const popularMovies = yield call(getAPI, popularMoviesPage1);
+    const request = yield call(
+      getAPI,
+      !query 
+        ? popularMoviesPage1 
+        : searchMovies
```
I'd take it outside the try block to a constant

MieczyslawFrontDev replied:
Great idea!

paulgrym commented on src/common/NoResultsPage/index.js:
```diff
@@ -3,10 +3,10 @@ import { NoResultsImage, Wrapper } from "./styled";
   return (
     <Section>
-      <Title> Sorry, there are no results for “Muan” </Title>
+      <Title>{`Sorry, there are no results for "${query[0].toUpperCase() + query.slice(1)}"`}</Title>
```
Consider making a function out of this and putting it in a separete file, because we will use it in MovieList and PersonList as well :)
```
const capitalize = (query) => query[0].toUpperCase() + query.slice(1);

paulgrym commented on src/common/Search/index.js:
```diff
@@ -0,0 +1,33 @@
+      value: target.value.trim() !== "" ? target.value : undefined,
+    });
+
+    target.value = "";
```
I think, it is disturbing when the query disappears after you stop typing 😁

MieczyslawFrontDev replied:
You're right.

paulgrym commented on src/features/movies/MovieList/popularMoviesSaga.js:
```diff
@@ -1,22 +1,37 @@
-    yield put(fetchMoviesError());
-  };
-};
+    yield put(fetchMoviesError);
```
We really need brackets here. Their absence causes problems with error handling.

MieczyslawFrontDev commented:
I made new changes to the search. I removed debounce and applied value. I added resetting the value after going to another subpage. The url now displays the page we are currently on. I have restored error and loading states in reducers. I also restored the redirect because after starting the application, the main page was empty and there should be a sub-page with movies. 
Thanks ;)

paulgrym commented on src/features/movies/MovieList/popularMoviesSaga.js:
```diff
@@ -1,22 +1,34 @@
+}
 
 export function* watchFetchPopularMovies() {
   yield takeLatest(fetchMovies.type, fetchPopularMoviesWorker);
```
Maybe debounce for 2s would be better

paulgrym reviewed (changes requested):
Search doesn't work properly - you can't type space after the first word

paulgrym commented on src/common/Search/index.js:
```diff
@@ -0,0 +1,42 @@
+    replaceQueryParameter([
+      {
+        key: queryParameters.search,
+        value: target.value.trim(),
```
This is the culprit of the search problem - you can't add a space, because it is trimmed immediately :(

paulgrym replied:
I see that value was `target.value.trim() !== "" ? target.value : undefined` in previous commits, let's get back to that :D

paulgrym commented on src/features/movies/MovieList/popularMoviesSlice.js:
```diff
@@ -23,7 +24,7 @@ const moviesSlice = createSlice({
     fetchMoviesSuccess: (state, { payload: movies }) => {
       state.loading = false;
-      state.error = false
+      state.error = false;
```
Good code cleaning 🧹

paulgrym commented on src/features/people/PersonList/popularPeopleSaga.js:
```diff
@@ -1,20 +1,27 @@
-    yield put(fetchPeopleSuccess(popularPeople));
+    yield delay(500);
+    const requestPeople = yield call(getAPI, urlPath);
+    console.log(requestPeople);
```
to remove :D

paulgrym commented on src/features/people/PersonList/popularPeopleSaga.js:
```diff
@@ -1,20 +1,27 @@
 
-function* fetchPopularPeopleWorker() {
-  const popularPeoplePage1 = `${APIPopularPeopleUrl}&page=1`;
+function* fetchPopularPeopleWorker( { payload: { query } }) {
+  const popularPeople = `${APIPopularPeopleUrl}&page=1`;
+  const searchPerson = `${APISearchPersonUrl}&query=${query}`;
```
We should pass page just as in MoviesSaga :D
